actioncable/lib/action_cable/subscription_adapter/compressor.rb

```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
+
+module ActionCable
+  module SubscriptionAdapter
+    class Compressor
+      def initialize(threshold: 1.kilobyte, compressor: ActiveSupport::Gzip)
+        @threshold = threshold
+        @compressor = compressor
+      end
+
+      def compress(data)
+        return "0#{data}" if data.size < @threshold
+
+        "1#{@compressor.compress(data)}"
+      end
+
+      def decompress(data)
+        return data.from(1) unless compressed?(data)
+
+        @compressor.decompress(data.from(1))
+      end
+
+      def compressed?(data)
+        data.start_with?("1")
```

Comment:
Are 0/1 sufficiently distinctive markers that they'll never trigger on real data?

Reply:
According to me, yes, because the prefix is always added, regardless of whether it was compressed or not. If the payload is "0", the payload to be broadcasted will be "00". Then the prefix is checked and removed

Reply:
It's more if the real payload had a 1, then decompression would fail.

And actually, on second thought, that could definitely be a problem. If you have a channel that sends ids, and we do, this is going to blow. Think we need a different marker.

Reply:
I don't quite understand this point. If you send an id with value "1", the actual payload to be broadcasted is "01". Then the prefix is removed. The same applies for compressed payloads. Just compress the payload without the prefix and then check in the decompress method if it has to decompress or not.

Anyway, maybe we could also use the same prefix of `ActiveSupport::Cache`, `\x78`, if you think it's better.

Reply:
> It's more if the real payload had a 1, then decompression would fail.

That only could happen during the upgrade roll out (when some instances still publish broadcasts without the compression flag). To prevent this (and potential _schema_) changes, it's worth using a custom prefix (kinda versioning mechanism).

Reply:
@palkan Right, do you have any suggestions to handle this?

Reply:
I see the following options here:

1) More sophisticated payload schema with the version bit: `| version (2) | compressed (1) | data |`. That would allow us not only handle legacy payloads but also allow us introduce new extensions in the future in a non-breaking way. What is the version then? Probably, something like the already mentioned `\x78` from the cache implementation (which is just `"x"`, right? so, it isn't really different form "1")

2) Implement a fallback: if we observe "1" and fail to decompress, fall back to the original value. However, that wouldn't work with payloads starting with "0". That leads us to the third option...

3) Just ignore. Action Cable pub/sub is ephemeral; if some messages would get lost (or failed to decompress), that wouldn't be the end of the world, since the communication channel is unreliable; so, we do not break any guarantees.

I would go with 3 for now; maybe, adding one more prefix to indicate the schema version, e.g., `"v1/<compressed>/<payload>"`.

Reply:
So in that case we should have an Action Cable setting that allows to configure the schema?

Reply:
> we should have an Action Cable setting that allows to configure the schema?

I don't think we need it; I meant internal version number (to handle framework-level changes), not smth like namespace